libsolidity/formal/BMC.cpp

```diff
@@ -951,18 +935,35 @@ void BMC::checkVerificationTarget(BMCVerificationTarget& _target)
 		case VerificationTargetType::Assert:
 			checkAssert(_target);
 			break;
+		case VerificationTargetType::ConstantCondition:
+			smtAssert(false, "Checks for constant condition are handled separately");
 		default:
-			solAssert(false, "");
+			smtAssert(false);
 	}
 }
 
-void BMC::checkConstantCondition(BMCVerificationTarget& _target)
+void BMC::checkIfConditionIsConstant(Expression const& _condition)
 {
+	// We ignore called functions here because they have specific input values.
+	// Also, expressions inside loop can have different values in different iterations.
+	if (!isRootFunction() || isInsideLoop())
+		return;
+
+	if (
+		!m_settings.targets.has(VerificationTargetType::ConstantCondition) ||
+		(m_currentContract && !shouldAnalyze(*m_currentContract))
```

Comment:
Following the change from #15880, I guess we don't want this check to be bypassed simply because it's in an unselected contract, right?

```suggestion
		(m_currentContract && !m_currentContract->canBeDeployed())
```

Reply:
Correct! The code has been updated.

Reply:
Actually, I wonder why we even have to explicitly check `shouldEncode()` here. This seems to be the only such check in `BMC.cpp`. Why aren't there more, for other targets? Are we missing some?